### DISCUSSION TITLE
Rule working in "*", not for language

### DISCUSSION THREAD

**Discussion by TonyGravagno** [zokugun/vscode-explicit-folding#105] [Q&A]
This works:
![image](https://github.com/zokugun/vscode-explicit-folding/assets/499473/097c0b55-394c-4630-be31-a58fee22c579)

```json
"editor.showFoldingControls": "always",
"editor.foldingStrategy": "auto",
"explicitFolding.rules": {
    "*": [
      {
        "begin": "return (",
        "end": ")"
      },
      {
        "foldLastLine": true
      }
    ],
```

This is not working with a .tsx file where the language is confirmed to be `typescriptreact` : 
![image](https://github.com/zokugun/vscode-explicit-folding/assets/499473/49da04cd-db7a-4cf4-9edc-4f6abe4285c9)

```json
"explicitFolding.rules": { 
...
    "[typescriptreact]": [
        {
          "begin": "return (",
          "end": ")"
        }
      ]
```

No joy when removing the brackets around `[typescriptreact]`.

Ideas?
Using latest published v0.23.1 with latest VSCode 1.79.1 over Win10.
I just noticed that my VSCode About shows Node 16.17.1, but the rest of the system is running 20.3.0. I'll look into that, don't think it's a factor, but mentioning just in case...

Thanks!

**Comment by TonyGravagno** [comment]
I have spent a Lot of time with this with `"explicitFolding.debug": true`.  
I was wrong.  
Do not use brackets around the language codes:   
 `typescriptreact` ![Do this](https://img.shields.io/badge/yup-darkgreen)  
`[typescriptreact]` ![Not this](https://img.shields.io/badge/nope-darkred)  

The "explicitFolding.rules" are processed from most specific to least specific. This is not dependent on the order in the settings.json file. That is, even if (in this example) the `*` rule is followed by all others, `typescriptreact` is still processed first and then the `*` catch-all is processed after.

So you don't need to duplicate rules in `*`, "just in case". But you also want to be careful to not put simple rules in the catch all that will apply to all file types. If you have a rule that only applies to code and not to text files, don't put it in the catch all, put the same rule in all of the code rule groups where it applies.

I might add other notes to this as I figure out more nuances.

**Reply by daiyam** [reply]
> Do not use brackets around the language codes:

I didn't catch the error when reading your post... Maybe I should support both.

Also you don't need to set `"foldLastLine": true` since it's the default value. (Also, it can be alone)

**Reply by TonyGravagno** [reply]
What I'm seeing in settings.json is that the root-level "settings" node does have [brackets] around language names, but the brackets don't work inside "explicitFolding.rules". Yes, there is confusion. It's not **necessary** to support both as long as this is clearly described in the doc. But yeah, I would support both. 😄

**Comment by TonyGravagno** [comment]
> Also you don't need to set `"foldLastLine": true` since it's the default value. (Also, it can be alone)

That one actually goes a bit deeper 😄 :

1) foldLastLine is not true by default when this extension is active and there are no rules for the current language.
2) This does not force the setting for all languages:

```json
   "explicitFolding.rules": { 
    "*": [
      {
        "foldLastLine": true
      }
    ],
```

3) This directive for a specific language is ignored, and yes, the default is true, but only if there is an active rule for the current language:

```json
    "typescriptreact": [
      {
        "foldLastLine": false
      },
```

4) Yes, this will Disable foldLastLine within a language-specific rule and only for that rule:

```json
      {
        "beginRegex": "aaa",
        "endRegex": "bbb",
        "nested": true,
        "foldLastLine": false
      },
```

I have tried many combinations, only noting a few here. I could take the time to test and document all behavior combinations but I lack the time to poke into the code to see if I can make changes. The most desirable setting is a cascade of priority from higher to lower, and it's almost like that, but not quite.

**Reply by daiyam** [reply]
> 1. foldLastLine is not true by default when this extension is active and there are no rules for the current language.

Ya you need rules. This extension only provides the folding ranges to VSCode. It doesn't control in anyway how the ranges are displayed.
